### PR TITLE
feat: Better CRD docs

### PR DESCRIPTION
**Pull request by fhennig**
# Description

Better CRD docs

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes
 
```[tasklist]
# Author
- [x] Changes are OpenShift compatible
- [x] CRD changes approved
- [x] Helm chart can be installed and deployed operator works
- [x] Integration tests passed (for non trivial changes)
- [x] Changes need to be "offline" compatible
```

```[tasklist]
# Reviewer
- [x] Code contains useful comments
- [x] (Integration-)Test cases added
- [x] Documentation added or updated
- [x] Changelog updated
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
```

### CODE REVIEW

**Comment by fhennig** [comment]
I consider this blocked on this ticket: https://github.com/stackabletech/spark-k8s-operator/issues/320

**Comment by adwk67** [comment]
Jenkins :green_circle: https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/spark-k8s-operator-it-custom/114/

File: Cargo.toml [hunk 1]
```diff
@@ -20,7 +20,7 @@ serde = { version = "1.0", features = ["derive"] }
 serde_json = "1.0"
 serde_yaml = "0.9"
 snafu = "0.7"
-stackable-operator = { git = "https://github.com/stackabletech/operator-rs.git", tag = "0.56.1" }
+stackable-operator = { git = "https://github.com/stackabletech/operator-rs.git", tag = "0.57.0" }
```
Comment (sbernauer) [inline Cargo.toml:23]: Nit: We could go all the way to 0.58.1 to pull in even better docs :)

File: rust/crd/src/lib.rs [hunk 2]
```diff
@@ -98,54 +104,97 @@ pub struct SparkApplicationStatus {
 )]
 #[serde(rename_all = "camelCase")]
 pub struct SparkApplicationSpec {
-    #[serde(default, skip_serializing_if = "Option::is_none")]
-    pub version: Option<String>,
-    #[serde(default, skip_serializing_if = "Option::is_none")]
-    pub mode: Option<String>,
+    /// Mode: cluster or client. Currently only cluster is supported.
+    pub mode: String,
```
Comment (sbernauer) [inline rust/crd/src/lib.rs:108]: Nit: While we already have a breaking change we could switch to an enum directly ;)

File: rust/crd/src/lib.rs [hunk 3]
```diff
@@ -98,54 +104,97 @@ pub struct SparkApplicationStatus {
...
+    /// The executor role specifies the configuration that, together with the driver pod template, is used by
+    /// Spark to create the executor pods.
     #[serde(default, skip_serializing_if = "Option::is_none")]
     pub executor: Option<RoleGroup<RoleConfigFragment>>,
-    #[serde(default, skip_serializing_if = "Option::is_none")]
-    pub stopped: Option<bool>,
+
+    /// A map of key/value strings that will be passed directly to spark-submit.
     #[serde(default, skip_serializing_if = "Option::is_none")]
     pub spark_conf: Option<HashMap<String, String>>,
```
Comment (sbernauer) [inline rust/crd/src/lib.rs:149]: Please also make mandatory. it will default to an empty hashmap anyway

File: rust/crd/src/lib.rs [hunk 4]
```diff
@@ -98,54 +104,97 @@ pub struct SparkApplicationStatus {
...
+
+    /// The job builds a spark-submit command, complete with arguments and referenced dependencies
+    /// such as templates, and passes it on to Spark.
     #[serde(default, skip_serializing_if = "Option::is_none")]
     pub job: Option<CommonConfiguration<SubmitConfigFragment>>,
+
+    /// The driver role specifies the configuration that, together with the driver pod template, is used by
+    /// Spark to create driver pods.
     #[serde(default, skip_serializing_if = "Option::is_none")]
     pub driver: Option<CommonConfiguration<RoleConfigFragment>>,
```
Comment (sbernauer) [inline rust/crd/src/lib.rs:140]: IMHO driver and executor roles should be mandatory, as you can not have a Spark job without those ;)

File: rust/crd/src/lib.rs [hunk 5]
```diff
@@ -98,54 +104,97 @@ pub struct SparkApplicationStatus {
...
     pub deps: Option<JobDependencies>,
+
+    /// Configure an S3 connection that the SparkApplication has access to.
+    /// Read more in the [Spark S3 usage guide](DOCS_BASE_URL_PLACEHOLDER/spark-k8s/usage-guide/s3).
     #[serde(default, skip_serializing_if = "Option::is_none")]
     pub s3connection: Option<S3ConnectionDef>,
+
+    /// Arguments passed directly to the job artifact.
     #[serde(default, skip_serializing_if = "Option::is_none")]
     pub args: Option<Vec<String>>,
```
Comment (sbernauer) [inline rust/crd/src/lib.rs:164]: Please also make mandatory. it will default to an empty vec anyway

File: rust/crd/src/lib.rs [hunk 6]
```diff
@@ -98,54 +104,97 @@ pub struct SparkApplicationStatus {
...
+
+    /// A list of volumes that can be made available to the job, driver or executors via their volume mounts.
     #[serde(default, skip_serializing_if = "Option::is_none")]
     pub volumes: Option<Vec<Volume>>,
```
Comment (sbernauer) [inline rust/crd/src/lib.rs:168]: Please also make mandatory. it will default to an empty vec anyway

File: rust/crd/src/lib.rs [hunk 7]
```diff
@@ -98,54 +104,97 @@ pub struct SparkApplicationStatus {
...
+
+    /// A list of environment variables that will be set in the job pod and the driver and executor
+    /// pod templates.
     #[serde(default, skip_serializing_if = "Option::is_none")]
```
Comment (sbernauer) [inline rust/crd/src/lib.rs:172]: Please also make mandatory. it will default to an empty vec anyway

File: rust/crd/src/lib.rs [hunk 8]
```diff
@@ -98,54 +104,97 @@ pub struct SparkApplicationStatus {
...
     pub packages: Option<Vec<String>>,
+
+    /// A list of repositories that is passed directly to `spark-submit`.
     #[serde(default, skip_serializing_if = "Option::is_none")]
     pub repositories: Option<Vec<String>>,
+
+    /// A list of excluded packages that is passed directly to `spark-submit`.
     #[serde(default, skip_serializing_if = "Option::is_none")]
     pub exclude_packages: Option<Vec<String>>,
```
Comment (sbernauer) [inline rust/crd/src/lib.rs:199]: For all 4 fields: Please also make mandatory. it will default to an empty vec anyway

File: CHANGELOG.md [hunk 9]
```diff
@@ -4,6 +4,12 @@ All notable changes to this project will be documented in this file.
 
 ## [Unreleased]
 
+### Added
+
+- Various documentation of the CRD ([#319]).
```
Comment (sbernauer) [inline CHANGELOG.md:9]: Please add something like `BREAKING: Made lots of attributes mandatory. While this change is technically breaking, your Spark job would have not worked before as well, as these attributes are needed`

File: rust/crd/src/lib.rs [hunk 10]
```diff
@@ -98,54 +104,97 @@ pub struct SparkApplicationStatus {
...
+
+    /// A map of key/value strings that will be passed directly to spark-submit.
     #[serde(default, skip_serializing_if = "Option::is_none")]
     pub spark_conf: Option<HashMap<String, String>>,
+
+    /// Job dependencies: a list of python packages that will be installed via pip, a list of packages
+    /// or repositories that is passed directly to spark-submit, or a list of excluded packages
+    /// (also passed directly to spark-submit).
     #[serde(default, skip_serializing_if = "Option::is_none")]
     pub deps: Option<JobDependencies>,
```
Comment (sbernauer) [inline rust/crd/src/lib.rs:155]: Nit: I would prefer making mandatory and providing a default using `#[serde(default]`

File: rust/crd/src/lib.rs [hunk 11]
```diff
@@ -98,54 +104,97 @@ pub struct SparkApplicationStatus {
...
+
+    /// A list of environment variables that will be set in the job pod and the driver and executor
+    /// pod templates.
     #[serde(default, skip_serializing_if = "Option::is_none")]
     pub env: Option<Vec<EnvVar>>,
+
+    /// The log file directory definition used by the Spark history server.
+    /// Currently only S3 buckets are supported.
     #[serde(default, skip_serializing_if = "Option::is_none")]
     pub log_file_directory: Option<LogFileDirectorySpec>,
```
Comment (sbernauer) [inline rust/crd/src/lib.rs:178]: Can a history server work without the directory? :thinking: If not please make mandatory as well

**Review by sbernauer — CHANGES_REQUESTED** [review]
Thanks for working on this and sorry for all these comments! But every `Option<>` we can get rid of IMHO makes the code better and less error-prone and prevents configuration errors. I'm also happy to do the proposed changes, in that case ping me :)

File: rust/crd/src/lib.rs (hunk 4 above)
Comment (adwk67) [reply rust/crd/src/lib.rs:140]: If the job/driver/executor roles are omitted completely, then Spark defaults to one driver and two executors and the job runs.

Comment (adwk67) [reply rust/crd/src/lib.rs:140]: This is about the simplest job that will run:
```
---
apiVersion: spark.stackable.tech/v1alpha1
kind: SparkApplication
metadata:
  name: spark-examples
spec:
  sparkImage:
    productVersion: "3.5.0"
    pullPolicy: IfNotPresent
  mode: cluster
  mainClass: org.apache.spark.examples.SparkALS
  mainApplicationFile: "local:///stackable/spark/examples/jars/spark-examples.jar"
```

File: rust/crd/src/lib.rs (hunk 3 above)
Comment (adwk67) [reply rust/crd/src/lib.rs]: See 27410e5.

File: rust/crd/src/lib.rs (hunk 5 above)
Comment (adwk67) [reply rust/crd/src/lib.rs]: See 27410e5.

File: rust/crd/src/lib.rs (hunk 6 above)
Comment (adwk67) [reply rust/crd/src/lib.rs]: See 27410e5.

File: rust/crd/src/lib.rs (hunk 7 above)
Comment (adwk67) [reply rust/crd/src/lib.rs]: See 27410e5.

File: rust/crd/src/lib.rs (hunk 8 above)
Comment (adwk67) [reply rust/crd/src/lib.rs]: See 27410e5.

File: CHANGELOG.md (hunk 9 above)
Comment (adwk67) [reply CHANGELOG.md:9]: See e4e6773.

File: rust/crd/src/lib.rs (hunk 4 above)
Comment (adwk67) [reply rust/crd/src/lib.rs:140]: I'd therefore be happy to leave them as optional - what do you think?

Comment (sbernauer) [reply rust/crd/src/lib.rs:140]: Yeah, should be fine (at the very least for now ;) )

**Review by sbernauer — APPROVED** [review]
It would be awesome if you could address the two remaining comments (also technically the changelog entry would be below `Changed` instead of `Added`)
I have to leave for vacation, please don't block the merge on me

**Comment by adwk67** [comment]
Thanks @sbernauer! I'm working on the last 2-3 points...

File: rust/crd/src/lib.rs (hunk 10 above)
Comment (adwk67) [reply rust/crd/src/lib.rs]: See 2a1bb77.

File: rust/crd/src/lib.rs (hunk 2 above)
Comment (adwk67) [reply rust/crd/src/lib.rs]: See 2a1bb77.

**Comment by adwk67** [comment]
Jenkins: https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/spark-k8s-operator-it-custom/115/

**Review by maltesander — APPROVED** [review]
LGTM!

**Pull request merged**